backend/lib/adapters/gateways/mongo/utils/mongo-adapter.ts

```diff
@@ -43,46 +45,52 @@ export class MongoCollectionAdapter<T> implements DocumentCollectionAdapter<T> {
     }
   }
 
-  public async paginatedFind(query: Pagination<T>): Promise<CamsPaginationResponse<T>> {
-    const mongoQuery = toMongoQuery<T>(query);
-    const countQuery = toMongoQuery<T>(query.values[0]);
+  public async paginate(
+    pipelineOrQuery: Pipeline | ConditionOrConjunction,
+    page: PaginationParameters = { offset: 0, limit: DEFAULT_SEARCH_LIMIT },
+  ): Promise<CamsPaginationResponse<T>> {
     try {
-      if (!isPagination(query)) {
-        throw new Error('Trying to paginate for a query that is not a pagination query');
-      }
+      const pipeline = isPipeline(pipelineOrQuery)
+        ? pipelineOrQuery
+        : QueryPipeline.pipeline(QueryPipeline.match(pipelineOrQuery));
 
-      const aggregationItem: CamsPaginationResponse<T> = {
-        metadata: { total: 0 },
-        data: [],
-      };
-
-      const aggregationResult = await this.collectionHumble.aggregate(mongoQuery);
-      aggregationItem.metadata.total = await this.collectionHumble.countDocuments(countQuery);
+      const includesPagination = pipeline.stages.reduce((acc, stage) => {
+        return acc || isPaginate(stage);
+      }, false);
 
-      for await (const result of aggregationResult) {
-        for (const doc of result.data) {
-          aggregationItem.data.push(doc as CamsItem<T>);
-        }
+      if (!includesPagination) {
+        pipeline.stages.push(QueryPipeline.paginate(page.offset, page.limit));
       }
+      const mongoAggregate = toMongoAggregate(pipeline);
+
+      const cursor = await this.collectionHumble.aggregate(mongoAggregate);
+      const result = await cursor.next();
 
-      return aggregationItem;
+      // NOTE: See the custom facet defined by `toMongoPaginatedFacet` in `mongo-aggregate-renderer.ts`.
+      // This stage is expected to be the final stage of any aggregate pipeline passed to `paginate`.
+      const paginationResult: CamsPaginationResponse<T> = {
+        metadata: result['metadata'] ? result['metadata'][0] : { total: 0 },
+        data: result['data'] ?? [],
+      };
+
+      return paginationResult;
     } catch (originalError) {
       throw this.handleError(
         originalError,
-        `Failed while querying with: ${JSON.stringify(query)}`,
-        { query },
+        `Failed while querying with: ${JSON.stringify(pipelineOrQuery)}`,
+        pipelineOrQuery,
```

Comment:
Here's one of those stringified objects we should get rid of. We're even passing the query as data, so totally unnecessary.